src/imap.cpp

```diff
@@ -1248,12 +1248,22 @@ void imap::parse_response(const string& response)
                     atom_state_ = atom_state_t::QUOTED;
                 }
                 else if (atom_state_ == atom_state_t::QUOTED)
-                    atom_state_ = atom_state_t::NONE;
+                {
+                    // The backslash and a double quote within an atom is the double quote only.
+                    if (token_list->back()->atom.back() != codec::BACKSLASH_CHAR)
+                        atom_state_ = atom_state_t::NONE;
+                    else
+                        token_list->back()->atom.back() = ch;
+                }
             }
             break;
 
             default:
             {
+                // Double backslash in an atom is translated to the single backslash.
+                if (ch == codec::BACKSLASH_CHAR && atom_state_ == atom_state_t::QUOTED && token_list->back()->atom.back() == codec::BACKSLASH_CHAR)
```

Comment:
thanks for your work!
i think it is good to check `token_list->back()->atom.size()`.
calling `std::string::back` on empty `std::string` object would cause undefined behaviour.
e.g. `"\abc"`
but there may be few cases like above example.
it's up to you.

Reply:
Hm, but in this case the atom is quoted, which means that the token contains at least the double quote character. Thus, calling `string::back()` is never on an empty string. Or I do not see something obvious?

Reply:
https://github.com/karastojko/mailio/blob/1124e1eb880d40a056bd4d7e1c835a935c0d02ec/src/imap.cpp#L1240-L1249
`imap::parse_response` method doesn't seem to put double quote character.
it may push back empty string token.

Reply:
`"\abc"` is the case.
1. receive `"` : enter into quoted state. `cur_token` is empty string
2. receive `\` : parser would execute `\` at first after entering into quoted state.
    - it may cause undefined behaviour since it doesn't check `token_list->back()->atom.size()` and call `token_list->back()->atom.back()`.

but it can work well fortunately. (not always)
for safe operation, it is a good to check `token_list->back()->atom.size()`, but if you don't think you need to do that, you don't need to do that.